core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -2814,17 +2814,41 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
       inode.setLastModificationTimeMs(opTimeMs);
     }
     if (options.getTtl() != null) {
-      Preconditions.checkArgument(inode.isFile(), PreconditionMessage.TTL_ONLY_FOR_FILE);
       long ttl = options.getTtl();
-      InodeFile file = (InodeFile) inode;
-      if (file.getTtl() != ttl) {
-        mTtlBuckets.remove(file);
-        file.setTtl(ttl);
-        mTtlBuckets.insert(file);
-        file.setLastModificationTimeMs(opTimeMs);
-        file.setTtlAction(options.getTtlAction());
+      if (inode.isDirectory()) {
+        InodeDirectory directory = (InodeDirectory) inode;
+        InodeDirectory parent = inodePath.getParentInodeDirectory();
+        if (parent != null) {
+          System.out.println("parent:" + parent.getTtl() + " " + ttl);
+          Preconditions.checkArgument(parent.getTtl() < 0 || ttl <= parent.getTtl(),
+              PreconditionMessage.TTL_FILE_SHOULD_LESS_THAN_PARENT_DIRECTORY);
+        }
+        if (directory.getTtl() != ttl) {
+          mTtlBuckets.remove(directory);
+          directory.setTtl(ttl);
+          mTtlBuckets.insert(directory);
+          directory.setLastModificationTimeMs(opTimeMs);
+          directory.setTtlAction(options.getTtlAction());
+        }
+        Set<Inode<?>> children = directory.getChildren();
+        for (Inode<?> child : children) {
+          AlluxioURI alluxioURI = new AlluxioURI(inodePath.getUri().getPath()
+              + "/" + child.getName());
+          setAttribute(alluxioURI, options);
```

Comment:
I think if we set TTL to a directory, recursive must be true to make sense --- otherwise what does it mean by expiring a directory but leaving all its children still in file system? 

For the parent TTL checking, actually I am fine with allowing children's TTL > parent's TTL. And when we deleting parent, we will delete children too. I will leave this decision to you. But I prefer to have less limitation --- when both solution A and B makes sense, I prefer the one requires less effort to explain to users.